### PR TITLE
Issue-3: Implement user profile page

### LINKED ISSUE #3: Implement user information updating (account settings page)
The front-end should have a way for a user to edit their info: email, username, and password. A user should also be able to delete their account from this page.

After the `delete` endpoint is hit, make sure to redirect the user to the signout page.

### PR DESCRIPTION
**Pull request by pyvelkov**
## Issue
closes #3 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [x] **UI**: Change which improves UI

## Description
Adds user settings/profile page which implements:
- updating user details (username, email, password)
- account deleting
- uploading new avatar (need to revisit with backend changes)

## Testing the PR
- cd web
- npm i
- npm run dev

## Checklist

- [ ] **Quality**: This PR builds and passes the respective npm run test and works locally
- [ ] **Tests**: This PR includes thorough tests
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Review by Thinbishop — CHANGES_REQUESTED** [review]
Testing unsuccessful. After changing password, I could not log in with either new or old password. I tried this with 2 users and it was the same both times.

File: web/public/luffy_placeholder_avatar.jpg
Comment (chrispinkney) [inline web/public/luffy_placeholder_avatar.jpg:1]: Do we need this?

File: web/src/app/dashboard/userprofile/page.tsx [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+								name="username"
+								value={userData.username}
+								onChange={handleUserDataChange}
+								label="Username"
+								type="text"
+								fullWidth
+								error={userData.username.length < 3 ? true : false}
+								helperText={
+									userData.username.length < 3
+										? 'Username has to be more than 3 chars.'
```
Comment (chrispinkney) [inline web/src/app/dashboard/userprofile/page.tsx:297]: In the code you do >= 3, so this message isn't accurate I think

File: web/src/app/dashboard/userprofile/page.tsx [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+								margin="dense"
+								name="password"
+								onChange={handleUserDataChange}
+								label="Password"
+								type="password"
+								fullWidth
+								error={userData.password.length < 6 ? true : false}
+								helperText={
+									userData.password.length < 6
+										? 'Password has to be more than 6 chars.'
```
Comment (chrispinkney) [inline web/src/app/dashboard/userprofile/page.tsx:328]: In the code you do >= 6, so this message isn't accurate I think

File: web/src/app/dashboard/userprofile/page.tsx [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+	const checkEmail = () => {
+		var re = /\S+@\S+\.\S+/;
+		return re.test(userData.email);
+	};
+
+	const handleSubmitData = async (e: React.FormEvent) => {
+		e.preventDefault();
+		const headers = { Authorization: `${token}` };
+		if (
+			userData.password.length >= 6 &&
```
Comment (chrispinkney) [inline web/src/app/dashboard/userprofile/page.tsx:129]: Didn't we add zod to the frontend? Can't we use that for checking password, username, and email?

File: web/src/app/dashboard/userprofile/page.tsx [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+		setDeleteEmail(e.target.value);
+	};
+
+	const handleUserDataChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+		const value = e.target.value;
+		setUserData({ ...userData, [e.target.name]: value });
+	};
+
+	const checkEmail = () => {
+		var re = /\S+@\S+\.\S+/;
```
Comment (chrispinkney) [inline web/src/app/dashboard/userprofile/page.tsx:121]: Can we use zod for this?

File: web/src/app/dashboard/userprofile/page.tsx [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+					User Profile
+				</Typography>
+				<Divider />
+			</Box>
+			<Box sx={{ flexGrow: 1 }}>
+				<Grid container spacing={2} sx={{ marginBottom: 4 }}>
+					<Grid item xs={1} sx={{ marginRight: 4 }}>
+						<Avatar
+							alt="PlaceHolder Avatar"
+							src="/luffy_placeholder_avatar.jpg"
```
Comment (chrispinkney) [inline web/src/app/dashboard/userprofile/page.tsx:178]: Can we use the initial avatar one instead?

**Review by chrispinkney — CHANGES_REQUESTED** [review]
Updating password doesn't work, but I think that's my fault, looks like after adding the encryption to the db I didn't encrypt the new password so they're all stored in plaintext and the user will never be able to sign in again after changing it (see Jon's message). I'll get on a fix for that soon.

Also is update avatar supposed to be working?

Also Also, everything is kind of misaligned due to the error messages:

image

File: web/public/luffy_placeholder_avatar.jpg
Comment (pyvelkov) [reply web/public/luffy_placeholder_avatar.jpg:1]: deleted it!

File: web/src/app/dashboard/userprofile/page.tsx (hunk 1 above)
Comment (pyvelkov) [reply web/src/app/dashboard/userprofile/page.tsx:402]: This is correct. Tested.

File: web/src/app/dashboard/userprofile/page.tsx (hunk 2 above)
Comment (pyvelkov) [reply web/src/app/dashboard/userprofile/page.tsx:433]: This is correct. Tested.

File: web/src/app/dashboard/userprofile/page.tsx [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+		setUserData({
+			username: whoamiJSON.username,
+			email: whoamiJSON.email,
+			password: '',
+		});
+	};
+
+	useEffect(() => {
+		getUser();
+	}, []);
```
Comment (chrispinkney) [inline web/src/app/dashboard/userprofile/page.tsx:81]: React Hook useEffect has missing dependencies: 'getAvatar' and 'getUser'. Either include them or remove the dependency array.eslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)

File: web/src/components/Header/AvatarLogo.tsx [hunk 7]
```diff
@@ -50,12 +51,45 @@ const AvatarLogo = () => {
...
 	useEffect(() => {
 		getUser();
+		getAvatar();
 	}, []);
 
+	useEffect(() => {
+		if (avatarURL == '') {
+			getAvatar();
+		}
+	});
```
Comment (chrispinkney) [inline web/src/components/Header/AvatarLogo.tsx:88]: without a dependency array doesnt this run forever if the condition passes?

**Review by Thinbishop — CHANGES_REQUESTED** [review]
The only major issue I've noticed is the `/download` being called due to the missing dependency of the useEffect (see @chrispinkney's review).

A couple suggestions I would give:
- The password in the user page is showing error without any interaction, it should start without an error until the user begins typing.
- We probably would benefit from a user context, this could contain the user id, username, avatar, etc. This would allow the Header avatar to update when the context changes, then we would only need one instance each of `/download` and `/whoami`.
- The user is required to enter their password to change their username or email. I'm not sure if this is intended, because the backend has all fields optional (incl. password) during user update.
- The page.tsx file is huge. It might be beneficial to break it up into smaller components to assist with readability and future debugging/extension.

**Comment by chrispinkney** [comment]
@pyvelkov Can you fix this PR to use the new util functions? (see [#106](https://github.com/chrispinkney/cinesync/pull/106/files))

File: web/src/components/UserProfile/DeleteUserProfile.tsx [hunk 8]
```diff
@@ -0,0 +1,100 @@
+'use client';
+import Button from '@mui/material/Button';
+import Dialog from '@mui/material/Dialog';
+import DialogActions from '@mui/material/DialogActions';
+import DialogContent from '@mui/material/DialogContent';
+import DialogContentText from '@mui/material/DialogContentText';
+import DialogTitle from '@mui/material/DialogTitle';
+import TextField from '@mui/material/TextField';
+import Grid from '@mui/material/Unstable_Grid2'; // Grid version 2
```
Comment (chrispinkney) [inline web/src/components/UserProfile/DeleteUserProfile.tsx:9]: remove comment

File: web/src/components/UserProfile/ProfileBottom.tsx [hunk 9]
```diff
@@ -0,0 +1,209 @@
+'use client';
+import { z } from 'zod';
+import Divider from '@mui/material/Divider';
+import Typography from '@mui/material/Typography';
+import Grid from '@mui/material/Unstable_Grid2'; // Grid version 2
```
Comment (chrispinkney) [inline web/src/components/UserProfile/ProfileBottom.tsx:5]: delete comment

File: web/src/components/UserProfile/ProfileTop.tsx [hunk 10]
```diff
@@ -0,0 +1,103 @@
+'use client';
+import { useUser } from '@/context/user.context';
+import Avatar from '@mui/material/Avatar';
+import Button from '@mui/material/Button';
+import Divider from '@mui/material/Divider';
+import Typography from '@mui/material/Typography';
+import Grid from '@mui/material/Unstable_Grid2'; // Grid version 2
```
Comment (chrispinkney) [inline web/src/components/UserProfile/ProfileTop.tsx:7]: delete comment

File: web/src/components/UserProfile/UploadAvatarImage.tsx [hunk 11]
```diff
@@ -0,0 +1,75 @@
+'use client';
+import Grid from '@mui/material/Unstable_Grid2'; // Grid version 2
```
Comment (chrispinkney) [inline web/src/components/UserProfile/UploadAvatarImage.tsx:2]: delete comment

File: web/src/components/UserProfile/UploadAvatarImage.tsx [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+			avatarFile = avatar.files?.[0];
+		}
+
+		if (avatarFile) {
+			formData.append('image', avatarFile);
+		}
+
+		try {
+			await fetch(
+				`${process.env.NEXT_PUBLIC_URL}/auth/upload` ||
```
Comment (chrispinkney) [inline web/src/components/UserProfile/UploadAvatarImage.tsx:42]: can you throw this into the new url file?

File: web/src/context/user.context.tsx [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+	useEffect,
+} from 'react';
+import { useGlobalContext } from './store';
+import {
+	deleteUserAvatar,
+	getUserAvatar,
+	getWhoAmI,
+} from '../utils/cinesync-api/fetch-user';
+
+// type User = {
```
Comment (chrispinkney) [inline web/src/context/user.context.tsx:18]: delete comment

File: web/src/context/user.context.tsx [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+
+const UserContext = createContext<UserContextType | undefined>(undefined);
+
+export const UserProvider = ({ children }: { children: ReactNode }) => {
+	const { token } = useGlobalContext();
+	const [user, setUser] = useState<User | null>(null);
+	const [avatar, setAvatar] = useState<string>('');
+
+	const getAvatar = useCallback(async () => {
+		// const { success, fetchResponseJson } = await getUserAvatar({
```
Comment (chrispinkney) [inline web/src/context/user.context.tsx:49]: delete comment

File: web/src/context/user.context.tsx [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+	}, [token]);
+
+	const refreshUser = useCallback(async () => {
+		const { success, fetchResponseJson } = await getWhoAmI({ token: token });
+		if (success && 'id' in fetchResponseJson) {
+			setUser(fetchResponseJson);
+		}
+	}, [token]);
+
+	// Come back to this later (used for header)
```
Comment (chrispinkney) [inline web/src/context/user.context.tsx:98]: delete comment

**Review by chrispinkney — CHANGES_REQUESTED** [review]

File: web/src/components/Header/AvatarLogo.tsx [hunk 16]
```diff
@@ -48,6 +51,11 @@ const AvatarLogo = () => {
 		getUser();
 	}, []);
 
+	// Come back to this later (see user.context)
```
Comment (Thinbishop) [inline web/src/components/Header/AvatarLogo.tsx:54]: This comment should be deleted. Maybe add as an issue?

**Review by Thinbishop — CHANGES_REQUESTED** [review]

**Review by chrispinkney — APPROVED** [review]

**Pull request merged**